### ISSUE TITLE
unsupported-operator for pytorch tensors

### ISSUE THREAD

**Issue by PetterS** [astral-sh/ty#908]
### Summary

Example:
```python
import torch
t = torch.zeros((10, 10))
t ** 2
```
ty:
```
error[unsupported-operator]: Operator `**` is unsupported between objects of type `Tensor` and `Literal[2]`
 --> ty_test.py:6:1
  |
5 | t = torch.zeros((10, 10))
6 | t ** 2
  | ^^^^^^
  |
info: rule `unsupported-operator` is enabled by default

Found 1 diagnostic
```

Other operators like `@` and `+` work.

### Version

ty 0.0.1-alpha.16 (c452e53a0 2025-07-25)

**Comment by sharkdp** [comment]
Thank you for reporting this.

The definition of `Tensor.__pow__` is here:

https://github.com/pytorch/pytorch/blob/f3913ea641d871f04fa2b6588a77f63efeeb9f10/torch/_tensor.py#L1084-L1092

At the moment, ty does *not* assume that a `Callable` attribute on a class is necessarily a descriptor. So when accessing `t.__pow__`, we assume that it does not bind the `self` argument. Consequently, calling `t.__pow__(2)` (which is more or less what happens when doing `t**2`), this leads to a `invalid-argument-type` error:

```
error[invalid-argument-type]: Argument is incorrect
 --> main.py:6:11
  |
5 | t = torch.zeros((10, 10))
6 | t.__pow__(2)
  |           ^ Expected `TensorBase`, found `Literal[2]`
```

This `invalid-argument-type` error is what causes the `unsupported-operator` on `t**2` (we should improve our diagnostics to make the root cause clear).

The behavior described above is a [known problem](https://discuss.python.org/t/when-should-we-assume-callable-types-are-method-descriptors/92938) in Python's type system. We are aware about the issue, and progress is currently tracked in https://github.com/astral-sh/ty/issues/491. So I'll close this as a duplicate, but mention your case in that ticket.

**Comment by PetterS** [comment]
Thanks for the explanation!
